lib/pengine/native.c

```diff
@@ -767,6 +805,13 @@ native_print(resource_t * rsc, const char *pre_text, long options, void *print_d
     common_print(rsc, pre_text, rsc_printable_id(rsc), node, options, print_data);
 }
 
+void
+pe__output_native(resource_t * rsc, long options, void * out)
+{
+    CRM_ASSERT(rsc->variant == pe_native);
+    pe__output_xml_native(rsc, options, out);
```

Comment:
I can do this. Do I understand correctly, that I should also remove the `void (*output) (pe_resource_t*, long, void*);` from the `struct resource_object_functions_s` in include/crm/pengine/pe_types.h:54 and call the messages method where required?

Reply:
The question is how do you know which message to call. Using the rsc->output() method, the right message will be called regardless of what kind of resource rsc is. If all the places output() is needed, we already know that, then it could make sense to just call the messages directly instead. I suppose another alternative would be to use a char* instead of a function in the resource object, so callers could just call out->message(out, rsc->msgtype, rsc)

Reply:
Now that you mention it, this might be a good idea. We don't even need a new object member, we could use crm_element_name(rsc->xml) to get "primitive", "clone", "group", "bundle".

Reply:
I removed the output member of `resource_object_functions_s` and `pe__output_native, pe__output_native, pe__output_native, and pe__output_native`. Now the messages are called instead. For example

```
out->message(out, crm_element_name(replica->ip->xml), options, replica->ip);
out->message(out, crm_element_name(replica->child->xml), options, replica->child);
out->message(out, crm_element_name(replica->container->xml), options, replica->container);
out->message(out, crm_element_name(replica->remote->xml), options, replica->remot
```